laplace/baselaplace.py

```diff
@@ -1012,8 +1054,489 @@ def sample(self, n_samples=100):
+        return f_mu.detach(), f_var.detach()
+
+    def _gp_posterior_variance(self, Js_star, X_star):
+        """
+        GP posterior variance: \\( k_{**} - K_{*M} (K_{MM}+ L_{MM}^{-1})^{-1} K_{M*}\\)
+
+        Parameters
+        ----------
+        Js_star : torch.Tensor
+            Jacobians of test data points
+        X_star : torch.Tensor
+            test data points \\(X \in \mathbb{R}^{N_{test} \\times C} \\)
+        """
+
+        K_star = self._kernel_star(Js_star, X_star)
+
+        K_M_star = []
+        for X_batch, _ in self.train_loader:
+            K_M_star_batch = self._kernel_batch_star(Js_star, X_batch)
+            K_M_star.append(K_M_star_batch)
+
+        f_var = K_star - self._build_K_star_M(K_M_star)
+        return f_var
+
+    def _build_K_star_M(self, K_M_star):
+        K_M_star = torch.cat(K_M_star, dim=1)
+        if self.diagonal_kernel:
+            prods = []
+            for c in range(self.n_outputs):
+                v = torch.squeeze(torch.linalg.solve(self.Sigma_inv[c], K_M_star[:, :, c].unsqueeze(2)), 2)
+                prod = torch.einsum('bm,bm->b', v, v)
+                prods.append(prod.unsqueeze(1))
+            prods = torch.cat(prods, dim=1)
+            return prods
+
+        else:
+            # in the reshape below we go from (N_test, M, C, C) to (N_test, M*C, C)
+            K_M_star = K_M_star.reshape(K_M_star.shape[0], -1, K_M_star.shape[-1])
+            v = torch.linalg.solve(self.Sigma_inv, K_M_star)
+            return torch.einsum('bcm,bcn->bmn', v, v)
+
+    def log_marginal_likelihood(self, prior_precision=None, sigma_noise=None):
+        """Compute the approximation to the log marginal likelihood subject
+
+        Requires that the Laplace approximation has been fit before.
+        The resulting torch.Tensor is differentiable in `prior_precision` and
+        `sigma_noise` if these have gradients enabled.
+        By passing `prior_precision` or `sigma_noise`, the current value is
+        overwritten. This is useful for iterating on the log marginal likelihood.
 
+        Parameters
+        ----------
+        prior_precision : torch.Tensor, optional
+            prior precision if should be changed from current `prior_precision` value
+        sigma_noise : [type], optional
+            observation noise standard deviation if should be changed
 
-class SoDLaplace(FunctionalLaplace):
-    pass
+        Returns
+        -------
+        log_marglik : torch.Tensor
+        """
+
+        # update prior precision (useful when iterating on marglik)
+        if prior_precision is not None:
+            self.prior_precision = prior_precision
+
+        # update sigma_noise (useful when iterating on marglik)
+        if sigma_noise is not None:
+            if self.likelihood != 'regression':
+                raise ValueError('Can only change sigma_noise for regression.')
+            self.sigma_noise = sigma_noise
+
+        self.fit(self.train_loader)
```

Comment:
indeed, refitting is not needed here

Reply:
changed the code so that refitting is not needed anymore here. note that due to this change now only isotropic priors are supported in `FunctionalLaplace`. if one uses per-parameter or per-layer priors, kernel matrices needs to be recomputed every time `prior_precision` is updated:

<img width="1072" alt="Screenshot 2022-12-20 at 14 37 57" src="https://user-images.githubusercontent.com/20437515/208680026-93518180-957c-435c-8030-ee632aea8e47.png">

added check `FucntionalLaplace` class